[primTorch] added ref for nn.functional.glu

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82819
* __->__ #82214
* #83524

Commit:
[primTorch] added ref for nn.functional.glu

[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

ngimel commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -559,3 +560,16 @@ def relu6(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
+@out_wrapper()
+def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
+    dim = utils.canonicalize_dims(a.ndim, dim)
+    if a.shape[dim] % 2:
```
use `check` here

ngimel commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -559,3 +560,16 @@ def relu6(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
+
+
+@register_decomposition(torch.ops.aten.glu)
+@out_wrapper()
```
you need type promotion wrapper here

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -21078,6 +21078,12 @@ def __init__(
+    PythonRefInfo(
+        "_refs.nn.functional.glu",
+        torch_opinfo_name="nn.functional.glu",
+        supports_nvfuser=False,
```
cc @IvanYashchuk, this seems like an easy pointwise operation, I'm surprised nvfuser doesn't support it

IvanYashchuk replied:
`torch.tensor_split` is not supported yet.
The error is `Node slice_in_dim_default with target prims.slice_in_dim.default does not support nvfuser`.
`prims.slice_in_dim` is used in the `torch._refs.tensor_split` implementation.

fdrocha commented on test/test_decomp.py:
```diff
@@ -159,6 +159,8 @@ def op_assert_ref(test_case, op, test_dtype, i, orig, decomp, ref, args, kwargs)
         (torch.float16, torch.ops.aten.native_batch_norm.default): 1e-5,
         (torch.bfloat16, torch.ops.aten.linalg_vector_norm.default): 1e-6,
         (torch.float16, torch.ops.aten.linalg_vector_norm.default): 1e-6,
+        (torch.float16, torch.ops.aten.glu.default): 1e-2,
+        (torch.bfloat16, torch.ops.aten.glu.default): 1e-2,
```
One thing I wanted to check explicitly is this. The tests don't pass unless I make the precision this low. It's only for {b,}float16 so maybe it's okay - what do you think @ngimel ?

Lezcano replied:
fwiw, this comes from the fact that the sigmoid is computed internally without casting yhe intermediate results to float32 as we do in TI

ngimel replied:
Oh you are right @Lezcano, we should fix that in the eager implementation

Lezcano replied:
I think eager is fine here, we should fix the ref. 
In the ref, we should first cast the inputs to float32, perform the computations in float32, and then cast the result back to whatever lower precision type we had.

Note that this behaviour is shared by all the TI implementations in eager, so this could very well be implemented as a decorator.

IvanYashchuk replied:
Just add type promotion wrapper here as was already suggested https://github.com/pytorch/pytorch/pull/82214#discussion_r939167379 and tests will pass without tolerance overrides.
```py
@elementwise_type_promotion_wrapper(
    type_promoting_args=("a",),
    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
)
```

Lezcano replied:
ag, that's it, thank you :)

fdrocha replied:
Yes, thanks!

ngimel replied:
Eager doesn't do type promotion to opmath https://github.com/pytorch/pytorch/blob/c3e67d8a8c1217c6c65c0539866e6c59ee2bd9a4/aten/src/ATen/native/cuda/UnarySpecialOpsKernel.cu#L154, whereas sigmoid ref does (all unary refs get wrapped in type promotion wrapper by default). glu ref also needs type promotion wrapper.

Lezcano replied:
We looked at it the other day and I completely missed that one! Thank you @ngimel!

fdrocha replied:
Thansk guys. @ngimel I think this should be ready to go, could you PTAL?

fdrocha replied:
Discussed it with @Lezcano and I will also fix up cpu and cuda implementations of sigmoid to do the type promotion for {,b}float16

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

Commit:
Update on "[primTorch] added ref for nn.functional.glu"




[ghstack-poisoned]

fdrocha commented:
Ok, now I think it really is ready. I added the type promotion logic to the cuda kernel. While doing it I noticed that sinc kernel was also missing it so I added it there too. @ngimel and @Lezcano could you PTAL?

ngimel reviewed (approved):
Awesome, looks good!

fdrocha commented:
@pytorchbot merge -g
